### ISSUE TITLE
Handling connection abort/failure (question)

### ISSUE THREAD

**Issue by tifroz** [swisspol/GCDWebServer#171]
Hi @swisspol & thank you for sharing GCDWebServer, it's a pretty amazing piece of software engineering :)

I am missing something however: I can't seem to be able to find a hook to handle a connection abort/fail. This is pretty important for my project where each connection can consume a substantial about of resources (I am using GCDWebServer as a proxy to stream a media hosted on another server - see code extract at the bottom).

In some circumstances (not always - I am unsure at this what the root cause is at this point), GCDWebServer will typically starts spitting out large numbers of errors like `Error while writing to socket 39: Broken pipe (32)` - presumably because the socket was disconnected by the client, but the proxy which  is blissfully unaware that the data has nowhere to go keeps sending data through. 

There must be a way to detect and handle connection abort / disconnect & free up the associated resources?

Note: I tried subclassing GCDWebServerConnection to at least detect the connection 'close' event, but the timing of the 'close' event does not really seem to correlate with the beginning of the 'Broken pipe' errors.

```
server.addHandlerForMethod("GET", pathRegex: pathRegex, requestClass: GCDWebServerRequest.classForCoder()) {
  var localResponse: GCDWebServerResponse!
  (gcdRequest:GCDWebServerRequest!, completion:GCDWebServerCompletionBlock!) -> Void in
  [...] // Asynchronous proxy client collects headers etc... into localResponse from remote server
  GCDWebServerStreamedResponse(contentType: mimeType, asyncStreamBlock: { (bodyReaderBlock: GCDWebServerBodyReaderCompletionBlock!) -> Void in
    [...] // Asynchronous proxy client collects data  from remote server
    bodyReaderBlock(data, error)
  }
  completion(localResponse)
}
``` 

Here's an extract from the logs (filtered out lines that didn't contain 'socket 25')

```
[275     ] 15:21:34.299 | general | Thread 5, name = (null) | [DEBUG] Did open connection on socket 25
[278     ] 15:21:34.303 | general | Thread 6, name = (null) | [DEBUG] Connection received 740 bytes on socket 25
[279     ] 15:21:34.305 | general | Thread 6, name = (null) | [DEBUG] Connection on socket 25 preflighting request "GET /webview/shouldInspectSegments" with 740 bytes body
[280     ] 15:21:34.305 | general | Thread 6, name = (null) | [DEBUG] Connection on socket 25 processing request "GET /webview/shouldInspectSegments" with 740 bytes body
[304     ] 15:21:34.324 | general | Thread 5, name = (null) | [DEBUG] Connection sent 206 bytes on socket 25
[305     ] 15:21:34.324 | general | Thread 7, name = (null) | [DEBUG] Connection sent 31 bytes on socket 25
[306     ] 15:21:34.332 | general | Thread 6, name = (null) | [DEBUG] Did close connection on socket 25
[322     ] 15:21:43.151 | general | Thread 5, name = (null) | [DEBUG] Did open connection on socket 25
[325     ] 15:21:43.153 | general | Thread 8, name = (null) | [DEBUG] Connection received 699 bytes on socket 25
[326     ] 15:21:43.155 | general | Thread 8, name = (null) | [DEBUG] Connection on socket 25 preflighting request "GET /proxy/http/cdn1.video.m.l3.swishly.com/80/default/videos/8/0/0/4/3/1/5/9/6/p360.mp4" with 699 bytes body
[327     ] 15:21:43.155 | general | Thread 8, name = (null) | [DEBUG] Connection on socket 25 processing request "GET /proxy/http/cdn1.video.m.l3.swishly.com/80/default/videos/8/0/0/4/3/1/5/9/6/p360.mp4" with 699 bytes body
[355     ] 15:21:43.259 | general | Thread 6, name = (null) | [DEBUG] Connection sent 489 bytes on socket 25
[358     ] 15:21:43.261 | general | Thread 7, name = (null) | [DEBUG] Connection sent 1 bytes on socket 25
[360     ] 15:21:43.262 | general | Thread 7, name = (null) | [DEBUG] Did close connection on socket 25
[362     ] 15:21:43.270 | general | Thread 7, name = (null) | [DEBUG] Did open connection on socket 25
[363     ] 15:21:43.271 | general | Thread 8, name = (null) | [DEBUG] Connection received 702 bytes on socket 25
[364     ] 15:21:43.273 | general | Thread 8, name = (null) | [DEBUG] Connection on socket 25 preflighting request "GET /proxy/http/cdn1.video.m.l3.swishly.com/80/default/videos/8/0/0/4/3/1/5/9/6/p360.mp4" with 702 bytes body
[365     ] 15:21:43.273 | general | Thread 8, name = (null) | [DEBUG] Connection on socket 25 processing request "GET /proxy/http/cdn1.video.m.l3.swishly.com/80/default/videos/8/0/0/4/3/1/5/9/6/p360.mp4" with 702 bytes body
[414     ] 15:21:43.325 | general | Thread 8, name = (null) | [DEBUG] Connection sent 499 bytes on socket 25
[416     ] 15:21:43.328 | general | Thread 8, name = (null) | [DEBUG] Connection sent 4344 bytes on socket 25
[418     ] 15:21:43.336 | general | Thread 7, name = (null) | [ERROR] Error while writing to socket 25: Broken pipe (32)
[421     ] 15:21:43.343 | general | Thread 6, name = (null) | [ERROR] Error while writing to socket 25: Broken pipe (32)
[422     ] 15:21:43.347 | general | Thread 7, name = (null) | [ERROR] Error while writing to socket 25: Broken pipe (32)
[424     ] 15:21:43.349 | general | Thread 7, name = (null) | [ERROR] Error while writing to socket 25: Broken pipe (32)
[...]
[7555    ] 15:23:42.648 | general | Thread 8, name = (null) | [ERROR] Error while writing to socket 25: Broken pipe (32)
[7558    ] 15:23:42.745 | general | Thread 7, name = (null) | [ERROR] Error while writing to socket 25: Broken pipe (32)
[7559    ] 15:23:42.745 | general | Thread 8, name = (null) | [ERROR] Error while writing to socket 25: Broken pipe (32)
[7569    ] 15:50:14.603 | general | Thread 8, name = (null) | [ERROR] Error while writing to socket 25: Broken pipe (32)
```

**Comment by swisspol** [comment]
> I can't seem to be able to find a hook to handle a connection abort/fail. 

There's none and it should not be needed if it's only to release resources. By definition the GCDWebServerConnection and GCDWebServerResponse will be de'alloced when not in use anymore (on either success or failure).

> but the proxy which is blissfully unaware that the data has nowhere to go keeps sending data through.

If there's an error while writing to the socket, I would expect the corresponding GCDWebServerConnection to just stop sending data and close itself (i.e. you should only see one error in the log). I'm not sure why it's not happening.

You would need to put a breakpoint here and trace what's happening in the code:
https://github.com/swisspol/GCDWebServer/blob/master/GCDWebServer/Core/GCDWebServerConnection.m#L261

In any case even if GCDWebServer keeps trying to send data, it's pretty much a no-op since the socket doesn't accept any. So things shouldn't be too bad and the connection will be closed soon afterwards anyway.

**Comment by tifroz** [comment]
Thank you so much for the help: that clarified a number of things. I was doing several things wrong, including calling `bodyReaderBlock` multiple times per invocation, and accidentally retaining `GCDWebServerResponse` ...turns out I wasn't clear on some of the subtleties of Swift closures either!

**Comment by swisspol** [comment]
Great to hear, I assume we can close this issue then :)

**Comment by SuperTango** [comment]
What about an async/async connection.  I've got a case where when getting a request on my GCDWebServer, I fire off a long running process that returns data back asynchronously.  I add an async handler using: 
```       
webServer.addDefaultHandler(forMethod: "GET", request: GCDWebServerRequest.self, asyncProcessBlock: { request, completionBlock in
```

and then when the handler is called, I created a GCDWebServerStreamedResponse later. 

```
let gcdWebServerResponse = GCDWebServerStreamedResponse(contentType: contentType, asyncStreamBlock: { bodyCompletionBlock in
```

And hang onto the bodyCompletionBlock so I can use it later. 

In the callback from my long running process, I call `bodyCompletionBlock(data, nil)`with whatever data.  However, there's no way for me to know if the GCDWebServer client has closed the connection (which they often do).  And thus my long running process keeps going.  I know that no data is being transferred, but I want to tell my long running process to stop when the GCDWebServer client closes the connection.  

Any ideas for this one?  

Thanks!
